File: examples/python/flat_index_example.py
"""Flat index -- brute-force exact search with 100% recall."""

import random
import uuid

from membrain import MembrainFlatIndex

DIMENSION = 128


def embed(seed: float) -> list[float]:
    random.seed(seed)
    return [random.gauss(0, 1) for _ in range(DIMENSION)]


def main() -> None:
    with MembrainFlatIndex(dimension=DIMENSION) as index:
        ids = [str(uuid.uuid4()) for _ in range(50)]
        for i, vector_id in enumerate(ids):
            index.add(vector_id, embed(seed=float(i)))

        print(f"Indexed {len(index)} vectors (dimension={index.dimension})")

        results = index.search(embed(seed=0.0), k=5)
        for result in results:
            print(f"  id={result.id[:8]}... score={result.score:.4f}")

        filtered = index.search_with_filter(embed(seed=0.0), k=3, allowed_ids=ids[:10])
        print(f"\nFiltered to 10 IDs: {len(filtered)} results")

        index.remove(ids[0])
        print(f"After remove: {len(index)} vectors")


if __name__ == "__main__":
    main()

File: examples/javascript/flat_index_example.mjs
/** Flat index -- brute-force exact search with 100% recall. */

import { randomUUID } from "node:crypto";
import { MembrainFlatIndex } from "membrain";

const DIMENSION = 128;

function embed(seed) {
  const out = [];
  let state = seed * 1000 + 1;
  for (let i = 0; i < DIMENSION; i++) {
    state = (state * 1103515245 + 12345) & 0x7fffffff;
    out.push((state / 0x7fffffff) * 2 - 1);
  }
  return out;
}

const index = new MembrainFlatIndex(DIMENSION);

try {
  const ids = Array.from({ length: 50 }, () => randomUUID());
  ids.forEach((id, i) => index.add(id, embed(i)));

  console.log(`Indexed ${index.len()} vectors (dimension=${index.dimension()})`);

  const results = index.search(embed(0), 5);
  for (const r of results) {
    console.log(`  id=${r.id.slice(0, 8)}... score=${r.score.toFixed(4)}`);
  }

  const filtered = index.searchWithFilter(embed(0), 3, ids.slice(0, 10));
  console.log(`\nFiltered to 10 IDs: ${filtered.length} results`);

  index.remove(ids[0]);
  console.log(`After remove: ${index.len()} vectors`);
} finally {
  index.close();
}
